indexing-hadoop/src/main/java/org/apache/druid/indexer/DetermineHashedPartitionsJob.java

```diff
@@ -129,10 +129,24 @@ public boolean run()
         JobHelper.writeJobIdToFile(config.getHadoopJobIdFileName(), groupByJob.getJobID().toString());
       }
 
-      if (!groupByJob.waitForCompletion(true)) {
-        log.error("Job failed: %s", groupByJob.getJobID());
-        failureCause = Utils.getFailureMessage(groupByJob, config.JSON_MAPPER);
-        return false;
+      try {
+        if (!groupByJob.waitForCompletion(true)) {
+          log.error("Job failed: %s", groupByJob.getJobID());
+          failureCause = Utils.getFailureMessage(groupByJob, config.JSON_MAPPER);
+          return false;
+        }
+      }
+      catch (IOException ioe) {
+        if (config.isUseYarnRMJobStatusFallback()) {
+          if (!Utils.checkAppSuccessFromYarnRM(groupByJob)) {
+            log.error(ioe, "Job failed: %s", groupByJob.getJobID());
+            failureCause =
+                "Could not retrieve job status from JobHistory server, YARN RM did not report job success either.";
+            return false;
+          }
```

Comment:
this pattern (line 132 - 146) appears to be repeated in many places, maybe extract it out into Utils.

Reply:
Added a new method `checkAppSuccessForJobIOException` that replaces these blocks